src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -995,42 +999,42 @@ export default class Blockchain extends Emittery.Typed<
       }
     };
 
-    let txHashCurrentlyProcessing: string = null;
-    const transactionHash = Data.from(transaction.hash()).toString();
+    const transactionHash = transaction.hash();
+    let txHashCurrentlyProcessing: Buffer = null;
 
-    const beforeTxListener = async (tx: Transaction) => {
-      txHashCurrentlyProcessing = Data.from(tx.hash()).toString();
-      if (txHashCurrentlyProcessing == transactionHash) {
+    const beforeTxListener = (tx: Transaction) => {
+      txHashCurrentlyProcessing = tx.hash();
+      if (txHashCurrentlyProcessing.equals(transactionHash)) {
         if (keys && contractAddress) {
-          keys.forEach(async key => {
-            // get the raw key using the hashed key
-            let rawKey = await this.#database.storageKeys.get(
-              Data.from(key).toString()
-            );
-
-            vm.stateManager.getContractStorage(
-              Address.from(contractAddress).toBuffer(),
-              rawKey,
-              (err: Error, result: Buffer) => {
-                if (err) {
-                  throw err;
+          const database = this.#database;
+          return Promise.all(
+            keys.map(async key => {
+              // get the raw key using the hashed key
+              const rawKey: Buffer = await database.storageKeys.get(key);
+
+              vm.stateManager.getContractStorage(
+                contractAddress,
+                rawKey,
+                (err: Error, result: Buffer) => {
+                  if (err) {
+                    throw err;
+                  }
+
+                  storage[Data.from(key, key.length).toString()] = {
+                    key: Data.from(rawKey, rawKey.length),
+                    value: Data.from(result, 32)
+                  };
```

Comment:
`Data` is pretty cool in that when it is JSON stringified it gets converted to RPC DATA. So there is no need to convert `.toJSON()` here.

By default, data remove leading `0`s, so if we want to keep them, which I believe we do in these cases, we need to specify the length.